git-protocol/src/fetch.rs

```diff
@@ -0,0 +1,493 @@
+    }
+
+    fn prepare_ls_refs(
+        &mut self,
+        _: &client::Capabilities,
+        args: &mut Vec<BString>,
+        _: &mut Vec<(&str, Option<&str>)>,
+    ) -> LsRefsAction {
+        for prefix in &self.opt.ref_prefixes {
+            args.push(format!("ref-prefix {}", prefix).into())
+        }
+
+        LsRefsAction::Continue
+    }
+
+    fn prepare_fetch(
+        &mut self,
+        _: git_transport::Protocol,
+        _: &client::Capabilities,
+        _: &mut Vec<(&str, Option<&str>)>,
+        refs: &[Ref],
+    ) -> Action {
+        self.out.refs = refs.into();
+
+        Action::Continue
+    }
+
+    fn negotiate(
+        &mut self,
+        refs: &[Ref],
+        args: &mut Arguments,
+        _: Option<&Response>,
+    ) -> io::Result<Action> {
+        for r in refs {
+            if let Some(WantHave { want, have }) = (self.filter_refs)(r)? {
+                args.want(want);
+                if let Some(oid) = have {
+                    args.have(oid);
+                }
+            }
+        }
+
+        for WantRef { name, have } in &self.opt.want_refs {
+            args.want_ref(name.as_bstr());
+            if let Some(oid) = have {
+                args.have(oid);
+            }
+        }
+
+        // send done, as we don't bother with further negotiation
+        Ok(Action::Cancel)
+    }
+}
+
+#[async_trait(?Send)]
+impl<F, P, O> Delegate for Fetch<F, P, O>
+where
+    F: Fn(&Ref) -> io::Result<Option<WantHave>>,
+    P: ReceivePack<Output = O>,
+{
+    async fn receive_pack(
+        &mut self,
+        pack: impl AsyncBufRead + Unpin + 'async_trait,
+        prog: impl Progress,
+        _: &[Ref],
+        resp: &Response,
+    ) -> io::Result<()> {
+        self.out.wanted_refs = resp
+            .wanted_refs()
+            .iter()
+            .map(|response::WantedRef { id, path }| Ref::Direct {
+                path: path.clone(),
+                object: *id,
+            })
+            .collect();
+        let out = self.receive_pack.receive_pack(pack, prog).await?;
+        self.out.pack = Some(out);
+
+        Ok(())
+    }
+}
+
+pub async fn fetch<F, P, O, R, W>(
+    opt: Options,
+    filter_refs: F,
+    receive_pack: P,
+    recv: R,
+    send: W,
+) -> io::Result<Outputs<O>>
+where
+    F: Fn(&Ref) -> io::Result<Option<WantHave>> + Send + 'static,
+    P: ReceivePack<Output = O> + Send + 'static,
+    O: Send + 'static,
+    R: AsyncRead + Unpin + Send + 'static,
+    W: AsyncWrite + Unpin + Send + 'static,
+{
+    use git_transport::{
+        client::git::{ConnectMode, Connection},
+        Protocol,
+    };
+
+    let conn = Connection::new(
+        recv,
+        send,
+        Protocol::V2,
+        opt.repo.clone(),
+        None::<(String, Option<u16>)>,
+        ConnectMode::Daemon,
+    );
+    blocking::unblock(move || {
+        let mut delegate = Fetch::new(opt, filter_refs, receive_pack);
+        future::block_on(git_protocol::fetch::fetch(
```

Comment:
Is this `block_on` inside `unblock` because the delegate can have blocking calls as per the documentation on `git_protocol::fetch::fetch`?

Reply:
I heard you like to block so I put some block in your unblock so you can block non-blocking!

Reply:
No seriously, this is basically a nested executor, such that blocking code (namely the machinery which writes the packfile to disk) can incrementally pull data from an `AsyncBufRead`.

In german, we call this "von hinten durch die Brust ins Auge".

Reply:
Ha, amazing idiom. Apparently the closest idiomatic English translation is to "do something ass backwards", which I don't really feel captures the implied complexity of the literal translation.

So this is because `Bundle::write_to_directory` requires `BufRead` and so we have to use `BlockOn` for that? I guess until we get async filesystem IO that's going to be an issue.

Reply:
I also saw:
> Going around your ass to get to your elbow

Which is appreciate a lot

Reply:
> So this is because Bundle::write_to_directory requires BufRead and so we have to use BlockOn for that? I guess until we get async filesystem IO that's going to be an issue.

Indeed, it's expected that this blocks. With `blocking` it shouldn't be an issue to run it within async code without blocking the executor. The only real question then is how the operation can be cancelled. Generally this is supposed to be done with these [little interrupt atomics](https://github.com/Byron/gitoxide/blob/38aca4076037a6f8288c2cf483f134ea16c328d5/gitoxide-core/src/pack/receive.rs#L201) that able to make the input reader fail. This could also be achieved by wrapping the input reader into a custom implementation, but `gitoxide` provides support to interrupt all potentially long running operations and may implement iterators on top of that. The latter play nicely with `blocking` as it will be able to stop ~~polling~~ calling `next()` on the iterator if the receiver of the computation is dropped.